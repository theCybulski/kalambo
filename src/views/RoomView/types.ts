export type RoomSettings = {
  roomId: string;
  adminId: string;
  gameLength: number;
}

export type Player = {
  id: string;
  name: string;
  score: number;
  isReady: boolean;
}

export type RoomRound = {
  isOn: boolean
  roundNo: number;
  drawingPlayerId: string;
  keyword: string;
  timer: number;
};

export type DrawingControls = {
  mode: string;
  strokeWidth: number;
  strokeColor: string;
}
