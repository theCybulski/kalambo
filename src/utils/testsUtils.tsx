import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'mobx-react';
import { ThemeProvider } from 'styled-components';

import playerStore from '__mocks__/store/PlayerStore';
import roomStore from '__mocks__/store/RoomStore';
import rootStore from '__mocks__/store/RootStore';

import defaultTheme from 'shared/styles/defaultTheme';
import lightTheme from 'shared/styles/lightTheme';

export const testTheme = { ...defaultTheme, ...lightTheme };

export const mountWithTheme = (component) =>
  mount(<ThemeProvider theme={{ ...testTheme }}>{component}</ThemeProvider>);

export const mountWithMockedStore = (component) =>
  mount(
    <Provider {...{ playerStore, roomStore, rootStore }}>
      <ThemeProvider theme={{ ...testTheme }}>{component}</ThemeProvider>
    </Provider>
  );
