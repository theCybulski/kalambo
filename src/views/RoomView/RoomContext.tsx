import React from 'react';
import { DrawingControls, Player, RoomRound, RoomSettings } from './types';

interface RoomContextProps {
  sockets: {
    chat: SocketIOClient.Socket;
    room: SocketIOClient.Socket;
  };
  settings: RoomSettings;
  setSettings?: React.Dispatch<React.SetStateAction<RoomSettings>>;
  localPlayer: Player;
  setLocalPlayer?: React.Dispatch<React.SetStateAction<Player>>;
  players: Player[];
  setPlayers?: React.Dispatch<React.SetStateAction<Player[]>>;
  round: RoomRound;
  setRound?: React.Dispatch<React.SetStateAction<RoomRound>>;
  drawingControls: DrawingControls;
  setDrawingControls?: React.Dispatch<React.SetStateAction<DrawingControls>>;
}

export const defaultValues = {
  sockets: {
    chat: null,
    room: null,
  },
  settings: {
    roomId: '',
    adminId: null,
    gameLength: 0,
  },
  localPlayer: {
    id: null,
    name: '',
    score: null,
    isReady: false,
  },
  players: [],
  round: {
    isOn: false,
    roundNo: null,
    drawingPlayerId: null,
    keyword: null,
    startedAt: null,
    length: 0,
  },
  drawingControls: {
    mode: 'brush',
    strokeWidth: 5,
    strokeColor: '#5405D4',
  },
};

export const RoomContext = React.createContext<RoomContextProps>(defaultValues);
