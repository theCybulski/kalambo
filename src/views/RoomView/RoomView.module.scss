@import "shared/styles/index";

.wrapper {
  display: flex;
  flex-direction: column;
  background-color: $blue-light;
  height: 100vh;
  padding: 25px 50px 40px;

  .grid {
    flex: 1;
    display: grid;
    grid-gap: 20px;
    grid-template-columns: repeat(12, 1fr);

    .flipChartWrapper {
      grid-column-start: 1;
      grid-column-end: 9;

      .cardFlipChart {
        position: relative;
        background-color: #fff;
        box-shadow: $shadow-primary;
        overflow: hidden;
        width: 100%;
        height: 100%;
        border-radius: 10px 20px 20px 10px;
      }
    }

    .chatWrapper {
      grid-column-start: 9;
      grid-column-end: 13;

      .cardChat {
        background-color: #fff;
        box-shadow: $shadow-primary;
        overflow: hidden;
        width: 100%;
        height: 100%;
        border-radius: 20px 10px 20px 20px;
      }
    }
  }
}
