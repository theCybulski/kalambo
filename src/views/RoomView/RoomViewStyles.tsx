import styled, { css } from 'styled-components';

export const Wrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    flex-direction: column;
    background-color: rgba(${theme.colors.secondaryFill}, 1);
    height: 100vh;
    padding: 25px 50px 40px;
  `}
`;

export const FlipChartWrapper = styled.div`
  grid-column-start: 1;
  grid-column-end: 9;
`;

export const SideElementsWrapper = styled.div`
  grid-column-start: 9;
  grid-column-end: 13;
`;

export const Card = styled.div`
  ${({ theme, corners }) => css`
    background-color: rgba(${theme.colors.primaryFill}, 1);
    box-shadow: ${theme.shadows.primary};
    border-radius: ${corners};
    overflow: hidden;
    width: 100%;
  `}
`;

export const CardFlipChart = styled(Card)`
  height: 100%;
`;

export const CardChat = styled(Card)`
  height: 70%;
`;

export const CardControls = styled(Card)`
  height: calc(30% - 20px);
  margin-top: 20px;
`;

export const Grid = styled.div`
  flex: 1;
  display: grid;
  grid-gap: 20px;
  grid-template-columns: repeat(12, 1fr);
`;
