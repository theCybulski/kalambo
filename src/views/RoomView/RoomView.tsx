import React, { useEffect, useRef } from 'react';
import { useHistory, Redirect } from 'react-router-dom';
import { parse } from 'query-string';
import { observer } from 'mobx-react';
import { useStores } from 'hooks/useStores';
import socket from 'api/api';

import FlipChart from 'components/FlipChart/FlipChart';
import DrawingControls from 'components/FlipChart/DrawingControls';
import TopBar from 'components/TopBar/TopBar';
import Chat from 'components/Chat/Chat';

import * as Styled from './RoomViewStyles';

const RoomView = observer(() => {
  const {
    playerStore: {
      localPlayer,
      localPlayer: { id: localPlayerId, roomNo: roomNumber, name: localPlayerName },
      setLocalPlayer,
    },
    roomStore: {
      drawingPlayerId,
      currRound: { isOn },
    },
  } = useStores();

  const isDrawing = localPlayerId === drawingPlayerId;

  const {
    location: { search },
  } = useHistory();

  const flipChart = useRef(null);

  useEffect(() => {
    if (socket) {
      const { id: roomNo } = parse(search);

      socket.emit(
        'joinRoom',
        {
          name: localPlayerName,
          roomNo,
        },
        ({ player, error }) => {
          error && console.log(new Error(error));
        }
      );

      return () => {
        socket.emit('disconnect', { localPlayer });
        // socket.off();
      };
    }
  }, [setLocalPlayer, localPlayerName, localPlayer, roomNumber, search]);

  const clearFlipChart = () => {
    flipChart.current.clearAll();
  };

  return (
    <Styled.Wrapper>
      {localPlayerId ? (
        <>
          <TopBar />
          <Styled.Grid>
            <Styled.FlipChartWrapper>
              <Styled.CardFlipChart corners="10px 20px 20px 10px">
                {socket && isOn && <FlipChart ref={flipChart} />}
              </Styled.CardFlipChart>
            </Styled.FlipChartWrapper>

            <Styled.SideElementsWrapper>
              <Styled.CardChat corners="20px 10px 20px 20px">{socket && <Chat />}</Styled.CardChat>
              <Styled.CardControls corners="20px 20px 10px 20px">
                {isDrawing && isOn && <DrawingControls onClear={clearFlipChart} />}
              </Styled.CardControls>
            </Styled.SideElementsWrapper>
          </Styled.Grid>
        </>
      ) : (
        <Redirect to="/" />
      )}
    </Styled.Wrapper>
  );
});

export default RoomView;
