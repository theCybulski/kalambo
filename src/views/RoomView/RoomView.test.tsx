import React from 'react';
import { Provider } from 'mobx-react';
import { shallow } from 'enzyme';
import { ThemeProvider } from 'styled-components';
import toJson from 'enzyme-to-json';

import playerStore from 'store/__mocks__/PlayerStore';
import roomStore from 'store/__mocks__/RoomStore';
import rootStore from 'store/__mocks__/RootStore';
import { testTheme } from 'utils/tests/testsUtils';

import RoomView from './RoomView';

describe(`${RoomView.name}`, () => {
  it('Matches snapshot', () => {
    const wrapper = shallow(
      <Provider {...{ playerStore, roomStore, rootStore }}>
        <ThemeProvider theme={{ ...testTheme }}>
          <RoomView />
        </ThemeProvider>
      </Provider>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
