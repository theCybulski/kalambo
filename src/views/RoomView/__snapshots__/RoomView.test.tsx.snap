// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`undefined Matches snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "playerStore": PlayerStore {},
      "roomStore": RoomStore {},
      "rootStore": RootStore {},
    }
  }
>
  <ThemeProvider
    theme={
      Object {
        "colors": Object {
          "additional": "50,50,50",
          "primaryColor": "113,31,218",
          "primaryFill": "255,255,255",
          "secondaryColor": "255,90,19",
          "secondaryFill": "249,248,253",
        },
        "deviceWidth": Object {
          "desktop": 1400,
          "desktopBig": 2000,
          "desktopSmall": 1260,
          "phone": 690,
          "phoneSmall": 370,
          "tablet": 1000,
        },
        "fonts": Object {
          "size": Object {
            "medium": "20px",
            "regular": "16px",
            "small": "12px",
          },
          "weight": Object {
            "bold": "700",
            "light": "300",
            "regular": "400",
          },
        },
        "shadows": Object {
          "primary": "0 10px 30px 0 rgba(84,37,201,0.06);",
        },
      }
    }
  >
    <Memo(wrappedComponent) />
  </ThemeProvider>
</ContextProvider>
`;
